src/evaluator.ts

```diff
@@ -982,6 +988,121 @@ class Evaluator {
+          new Promise<void>((_, reject) => {
+            const timeoutId = setTimeout(() => {
+              // Abort any ongoing requests
+              abortController.abort();
+
+              // If the provider has a cleanup method, call it
+              if (typeof evalStep.provider.cleanup === 'function') {
+                try {
+                  evalStep.provider.cleanup();
+                } catch (cleanupErr) {
+                  logger.warn(`Error during provider cleanup: ${cleanupErr}`);
+                }
+              }
+
+              reject(new Error(`Evaluation timed out after ${timeoutMs}ms`));
+
+              // Clear the timeout to prevent memory leaks
+              clearTimeout(timeoutId);
+            }, timeoutMs);
+          }),
+        ]);
+      } catch (error) {
+        // Create and add an error result for timeout
+        const timeoutResult = {
+          provider: {
+            id: evalStep.provider.id(),
+            label: evalStep.provider.label,
+            config: evalStep.provider.config,
+          },
+          prompt: {
+            raw: evalStep.prompt.raw,
+            label: evalStep.prompt.label,
+            config: evalStep.prompt.config,
+          },
+          vars: evalStep.test.vars || {},
+          error: `Evaluation timed out after ${timeoutMs}ms: ${String(error)}`,
+          success: false,
+          failureReason: ResultFailureReason.ERROR, // Using ERROR for timeouts
+          score: 0,
+          namedScores: {},
+          latencyMs: timeoutMs,
+          promptIdx: evalStep.promptIdx,
+          testIdx: evalStep.testIdx,
+          testCase: evalStep.test,
+          promptId: evalStep.prompt.id || '',
+        };
+
+        // Add the timeout result to the evaluation record
+        await this.evalRecord.addResult(timeoutResult);
+
+        // Update stats
+        this.stats.errors++;
+
+        // Update prompt metrics
+        const { promptIdx } = timeoutResult;
+        const metrics = prompts[promptIdx].metrics;
+        if (metrics) {
+          metrics.testErrorCount += 1;
+          metrics.totalLatencyMs += timeoutMs;
+        }
+
+        // Progress callback
+        if (options.progressCallback) {
+          options.progressCallback(
+            this.evalRecord.resultsCount,
+            runEvalOptions.length,
+            typeof index === 'number' ? index : 0,
+            evalStep,
+            metrics || {
+              score: 0,
+              testPassCount: 0,
+              testFailCount: 0,
+              testErrorCount: 1,
+              assertPassCount: 0,
+              assertFailCount: 0,
+              totalLatencyMs: timeoutMs,
+              tokenUsage: {
+                total: 0,
+                prompt: 0,
+                completion: 0,
+                cached: 0,
+                numRequests: 0,
+              },
+              namedScores: {},
+              namedScoresCount: {},
+              cost: 0,
+            },
```

Comment:
thought: this seems long and potentially something that could be reused. I wonder if we should wrap metrics in a class and have this be the default for empty metrics